### PR TITLE
[ADD] doc: coding guidelines

### PR DESCRIPTION
**Pull request by mart-e**
Write guidelines on how to write secure code and avoid commons
mistakes.

Update outdated information and move existing security guidelines to
the same section.

Based on [security checklist](https://drive.google.com/file/d/1-BKF6kcVO-vQapH5dpSeUI7_40wDHtLa/view)

### CODE REVIEW

File: doc/reference/security.rst [hunk 1]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+As a developer, it is important to understand the security mechanisms and avoid
+common mistakes leading to unsecure code.
+
+Unsafe Public Methods
+---------------------
+
+Any public method can be execute via a :ref:`RPC call
+<webservices/odoo/calling_methods>` with the chosen parameters. The methods
+starting with a `_` are considered as private in Odoo and not callable from
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
starting with a ``_`` are considered as private in Odoo and not callable from
```

File: doc/reference/security.rst [hunk 2]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+- The problem with query parameters (http://initd.org/psycopg/docs/usage.html#the-problem-with-the-query-parameters)
+- How to pass parameters with psycopg2 (http://initd.org/psycopg/docs/usage.html#passing-parameters-to-sql-queries)
+- Advanced parameter types (http://initd.org/psycopg/docs/usage.html#adaptation-of-python-values-to-sql-types)
+
+Unescaped field content
+-----------------------
+
+When rendering content using JavaScript and XML, one may be tempted to use
+a `t-raw` to display rich-text content. This should be avoided as a frequent
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
a ``t-raw`` to display rich-text content. This should be avoided as a frequent
```

File: doc/reference/security.rst [hunk 3]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+    auction_lots_ids = self.search([('auction_id','in',ids), ('state','=','draft'), ('obj_price','>',0)])
+
+
+SQL injections
+--------------
+Care must be taken not to introduce SQL injections vulnerabilities when using
+manual SQL queries. The vulnerability is present when user input is either
+incorrectly filtered or badly quoted, allowing an attacker to introduce
+undesirable clauses to a SQL query (such as circumventing filters or
+executing `UPDATE` or `DELETE` commands).
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
executing ``UPDATE`` or ``DELETE`` commands).
```

File: doc/reference/security.rst [hunk 4]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+-----------------------
+
+When rendering content using JavaScript and XML, one may be tempted to use
+a `t-raw` to display rich-text content. This should be avoided as a frequent
+XSS_ vector.
+
+.. _XSS: https://en.wikipedia.org/wiki/Cross-site_scripting
+
+It is very hard to control the integrity of the data from the computation until
+the final integration in the browser DOM. A `t-raw` that is correctly escaped at
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
the final integration in the browser DOM. A ``t-raw`` that is correctly escaped at
```

File: doc/reference/security.rst [hunk 5]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+    })
+
+.. code-block:: xml
+
+    <div t-name="insecure_template">
+        <div id="information-bar"><t t-raw="info_message" /></div>
+    </div>
+
+The above code may feel safe as the message content is controlled but is a bad
+pratice that may lead to unexpected security vulnerabilities once this code
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
practice that may lead to unexpected security vulnerabilities once this code
```

File: doc/reference/security.rst [hunk 6]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+.. code-block:: javascript
+
+    // XSS possible with unescaped user provided content !
+    QWeb.render('insecure_template', {
+        info_message: "You have an <strong>important</strong> notification on " \
+            + "the product <strong>" + product.name + "</strong>",
+    })
+
+While formating the template differently would prevent such vulnerabilities.
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
While formatting the template differently would prevent such vulnerabilities.
```

File: doc/reference/security.rst [hunk 7]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+.. code-block:: css
+
+    .subject {
+        font-weight: bold;
+    }
+
+Evaluating content
+------------------
+Some may want to `eval` to parse user provided content. Using `eval` should be
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
Some may want to ``eval`` to parse user provided content. Using ``eval`` should be
```

File: doc/reference/security.rst [hunk 8]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+avoided at all cost. A safer, sandboxed, method :class:`~odoo.tools.safe_eval`
+can be used instead but still gives tremendous capabilities to the user running
+it and must be reserved for trusted priviledged users only as it breaks the
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
it and must be reserved for trusted privileged users only as it breaks the
```

File: doc/reference/security.rst [hunk 9]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+    from odoo.tools import safe_eval
+    domain = safe_eval(self.filter_domain)
+    return self.search(domain)
+
+    # good
+    from ast import literal_eval
+    domain = literal_eval(self.filter_domain)
+    return self.search(domain)
+
+Parsing content does not need `eval`
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
Parsing content does not need ``eval``
```

File: doc/reference/security.rst [hunk 10]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+Javascript  int, float, etc.  parseInt(), parseFloat()
+Python      dict              json.loads(), ast.literal_eval()
+Javascript  object            JSON.parse()
+==========  ================  ================================
+
+Accessing object attributes
+---------------------------
+
+If the values of a record needs to be retrieved or modified dynamically, one may
+want to use the `getattr` and `setattr` methods.
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
want to use the ``getattr`` and ``setattr`` methods.
```

File: doc/reference/security.rst [hunk 11]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+.. code-block:: python
+
+    # unsafe retrieval of a field value
+    def _get_state_value(self, res_id, state_field):
+        record = self.sudo().browse(res_id)
+        return getattr(record, state_field, False)
+
+This code is however not safe as allow to access any property of the record,
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
This code is however not safe as it allows to access any property of the record,
```

File: doc/reference/security.rst [hunk 12]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+including private attributes or methods.
+
+The `__getitem__` of a recordset has been defined and accessing a dynamic field
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
The ``__getitem__`` of a recordset has been defined and accessing a dynamic field
```

File: doc/reference/security.rst [hunk 13]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+.. code-block:: python
+
+    # better retrieval of a field value
+    def _get_state_value(self, res_id, state_field):
+        record = self.sudo().browse(res_id)
+        return record[state_field]
+
+The above method is obviously still too optimistic and additional verifications
+on the record id and field value must be checked.
```
Comment (AntoineVDV) [inline doc/reference/security.rst]:
```suggestion
on the record id and field value must be done.
```

File: doc/reference/security.rst [hunk 14]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+        info_message: "You have an <strong>important</strong> notification on " \
+            + "the product <strong>" + product.name + "</strong>",
+    })
+
+While formatting the template differently would prevent such vulnerabilities.
+
+.. code-block:: javascript
+
+    QWeb.render('secure_template', {
+        message: "You have an important notification on the product:",
```
Comment (william-andre) [inline doc/reference/security.rst]: `important` is not rendered in bold like in the insecure template though

**Comment by pedrobaeza** [comment]
@mart-e do you want to continue this or should be closed?

**Comment by mart-e** [comment]
No no, stil to be merged ;-)

File: doc/reference/security.rst [hunk 15]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+Security Pitfalls
+=================
+
+As a developer, it is important to understand the security mechanisms and avoid
+common mistakes leading to unsecure code.
+
+Unsafe Public Methods
+---------------------
+
+Any public method can be execute via a :ref:`RPC call
```
Comment (xmo-odoo) [inline doc/reference/security.rst:111]:
```suggestion
Any public method can be executed via a :ref:`RPC call
```

File: doc/reference/security.rst [hunk 16]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+.. _time module: https://docs.python.org/3/library/time.html
+
+
+.. _reference/security/pitfalls:
+
+Security Pitfalls
+=================
+
+As a developer, it is important to understand the security mechanisms and avoid
+common mistakes leading to unsecure code.
```
Comment (xmo-odoo) [inline doc/reference/security.rst:106]:
```suggestion
common mistakes leading to insecure code.
```

File: doc/reference/security.rst [hunk 17]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+Unsafe Public Methods
+---------------------
+
+Any public method can be execute via a :ref:`RPC call
+<webservices/odoo/calling_methods>` with the chosen parameters. The methods
+starting with a ``_`` are considered as private in Odoo and not callable from
```
Comment (xmo-odoo) [inline doc/reference/security.rst:113]:
```suggestion
starting with a ``_`` are not callable from
```
they're not private in the sense readers would usually understand: they're often very much intended to be used or overridden, just not callable over RPC.

File: doc/reference/security.rst (hunk 15 above)
Comment (xmo-odoo) [reply doc/reference/security.rst:111]: Also maybe mention that there are no ACL checks for public methods, ACLs are only checked when specifically performing CRUD operations? So they're callable with essentially no access rights, possibly none at all using the webclient's RPC thing?

File: doc/reference/security.rst [hunk 18]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+        self.sudo().write({"state": new_state})
+
+Making a method private is obviously not enough and care must be taken to use it
+properly.
+
+Bypassing the ORM
+-----------------
+You should never use the database cursor directly when the ORM can do the same
+thing! By doing so you are bypassing all the ORM features, possibly the
+transactions, access rights and so on.
```
Comment (xmo-odoo) [inline doc/reference/security.rst:137]: nit: probably not the transactions as that's an sql-level thing, however it's bypassing a bunch of automated behaviours like translations, automatic fields, `active` field handling (though sometimes these could be considered misfeatures, YMMV)

File: doc/reference/security.rst [hunk 19]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+    self.env.cr.execute('SELECT id FROM auction_lots WHERE auction_id in (' + ','.join(map(str, ids))+') AND state=%s AND obj_price > 0', ('draft',))
+    auction_lots_ids = [x[0] for x in self.env.cr.fetchall()]
+
+    # no injection, but still wrong
+    self.env.cr.execute('SELECT id FROM auction_lots WHERE auction_id in %s '\
+               'AND state=%s AND obj_price > 0', (tuple(ids), 'draft',))
+    auction_lots_ids = [x[0] for x in self.env.cr.fetchall()]
+
+    # better
+    auction_lots_ids = self.search([('auction_id','in',ids), ('state','=','draft'), ('obj_price','>',0)])
```
Comment (xmo-odoo) [inline doc/reference/security.rst:154]: mebbe the code blocks could be wrapped in admonitions to outline them as dangerous / warning / good?

File: doc/reference/security.rst [hunk 20]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+
+SQL injections
+--------------
```
Comment (xmo-odoo) [inline doc/reference/security.rst:158]: Could be a subsection of the previous:

### YOU SHALL NOT BYPASS THE ORM

#### if you really need to, here's what to do

File: doc/reference/security.rst [hunk 21]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+This is very important, so please be careful also when refactoring, and most
+importantly do not copy these patterns!
+
+Here is a memorable example to help you remember what the issue is about (but
+do not copy the code there). Before continuing, please be sure to read the
+online documentation of pyscopg2 to learn of to use it properly:
+
+- The problem with query parameters (http://initd.org/psycopg/docs/usage.html#the-problem-with-the-query-parameters)
+- How to pass parameters with psycopg2 (http://initd.org/psycopg/docs/usage.html#passing-parameters-to-sql-queries)
+- Advanced parameter types (http://initd.org/psycopg/docs/usage.html#adaptation-of-python-values-to-sql-types)
```
Comment (xmo-odoo) [inline doc/reference/security.rst:197]: Would probably deserve a "dynamic SQL" subsection, as that'd be the most common real need for SQL (the other generally being performances), and mention of https://www.psycopg.org/docs/sql.html

File: doc/reference/security.rst [hunk 22]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+Unescaped field content
+-----------------------
+
+When rendering content using JavaScript and XML, one may be tempted to use
+a ``t-raw`` to display rich-text content. This should be avoided as a frequent
+XSS_ vector.
+
+.. _XSS: https://en.wikipedia.org/wiki/Cross-site_scripting
```
Comment (xmo-odoo) [inline doc/reference/security.rst:206]:
```suggestion
`XSS <https://en.wikipedia.org/wiki/Cross-site_scripting>`_ vector.
```
the footnote-y version is useful when you want to link multiple references.

File: doc/reference/security.rst [hunk 23]
```diff
@@ -93,8 +91,233 @@ access to the field:
...
+
+Parsing content does not need ``eval``
+
+==========  ================  ================================
+Language    Data type         Suitable parser
+==========  ================  ================================
+Python      int, float, etc.  int(), float()
+Javascript  int, float, etc.  parseInt(), parseFloat()
+Python      dict              json.loads(), ast.literal_eval()
+Javascript  object            JSON.parse()
```
Comment (xmo-odoo) [inline doc/reference/security.rst:293]: also list / array.

**Review by xmo-odoo — COMMENTED** [review]
Typos and possible improvements.

File: doc/reference/security.rst (hunk 20 above)
Comment (william-andre) [reply doc/reference/security.rst:158]: Meanwhile in the `upgrade` repo...

**Comment by mart-e** [comment]
robodoo r+

**Pull request closed without merging**